keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md

```diff
@@ -275,20 +275,48 @@ with a metric.
 
 There could be a large volume of requests for a specific resource which might result in the identified apiserver being unable to serve the proxied requests. This scenario should not occur too frequently, since resource types which have large request volume should not be added or removed during an upgrade -- that would cause other problems, too.
 
-TODO: security / cert stuff.
+We should ensure a definite way for the request flow to complete rather than continuing on indefinitely by proxying the request over and over again.
+
+We need to watch out for inadvertantly putting apiserver IP/endpoint information and the trust bundle control (used to authenticate server while proxying) in user hands (by exposing this information in REST APIs)
 
 ## Design Details
 
-TODO: explanation of how the handler will determine a request is for a resource
-that should be proxied.
+### Aggregation Layer
+
+1. A new filter will be added to the [handler chain] of the aggregation layer. This filter will maintain an internal map with the key being the group-version-resource and the value being a list of server IDs of apiservers that are capable of serving that group-version-resource.
+
+    1. This internal map is populated using an informer for StorageVersion objects. An event handler will be added for this informer that will get the apiserver ID of the requested group-version-resource and update the internal map accordingly.
+
+2. This filter will pass on the request to the next handler in the local aggregator chain, if:
+    1. It is a non resource request
+    2. No StorageVersion was retrieved for it, meaning the request is for an aggregated API or for a custom resource
+    3. If the local apiserver ID is found in the list of serviceable-by server IDs from the internal map
+
+3. If the local apiserver ID is not found in the list of serviceable-by server IDs, a random apiserver ID will be selected from the retrieved list and the request will be proxied to this apiserver.
+
+4. If there is no apiserver ID retrieved for the requested GVR, we will serve 503 with error `No relevant API server found for the requested GVR <group_version_resource>`
```

Comment:
But if the storageversion for the gvr was found, could the missing apiserver ID be because of some delay in updating the records? Is it correct to serve 404 still?

Reply:
Got it. For the case when there is no storage version found, we pass the request to the next handler in local apiserver, since it could be a request for an aggregated API or for a CR

But in the case of not finding any apiserver, should we still pass on the request as it is to the next handler or short circuit to 404?

Reply:
I think it will get a 404 either way, so it doesn't really matter.

(I assume we already detected CRs and AAs and let them proceed down the handler chain)

Reply:
Okay, updated text to serve 404 in this case. Thanks for correcting my understanding!